iree/base/api.h

```diff
@@ -274,45 +476,203 @@ typedef uintptr_t iree_status_t;
+// |file| and |line| should be populated with __FILE__ and __LINE__ at the call
+// site and an optional string |message| may be provided.
+//
+// The status will be allocated using the default system allocator and must be
+// freed using either iree_status_free or iree_status_ignore.
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_allocate(iree_status_code_t code, const char* file, uint32_t line,
+                     iree_string_view_t message);
+
+// Allocates a new status instance for a failing error |code| and annotates it
+// with a printf-style format string. Roughly equivalent (though more efficient)
+// than iree_status_allocate + iree_status_annotate_f.
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+    IREE_PRINTF_ATTRIBUTE(4, 5)
+        iree_status_allocate_f(iree_status_code_t code, const char* file,
+                               uint32_t line, const char* format, ...);
+
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_allocate_vf(iree_status_code_t code, const char* file,
+                        uint32_t line, const char* format, va_list varargs_0,
+                        va_list varargs_1);
+
+// Clones |status| into a new status instance.
+// No payloads, if present, will be cloned.
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_clone(iree_status_t status);
+
+// Frees |status| if it has any associated storage.
+IREE_API_EXPORT void IREE_API_CALL iree_status_free(iree_status_t status);
+
+// Ignores |status| regardless of its value and frees any associated payloads.
+// Returns an OK status that can be used when chaining.
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_status_ignore(iree_status_t status);
+
+// Consumes the |status| by freeing its storage and returning its code.
+IREE_API_EXPORT iree_status_code_t IREE_API_CALL
+iree_status_consume_code(iree_status_t status);
+
+// Annotates a status message with the given constant string message.
+// Ignored if |base_status| is OK.
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_annotate(iree_status_t base_status, iree_string_view_t message);
+
+// Annotates a status message with the given printf-style message.
+// Ignored if |base_status| is OK.
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+    IREE_PRINTF_ATTRIBUTE(2, 3)
+        iree_status_annotate_f(iree_status_t base_status, const char* format,
+                               ...);
+
+IREE_API_EXPORT IREE_MUST_USE_RESULT iree_status_t IREE_API_CALL
+iree_status_annotate_vf(iree_status_t base_status, const char* format,
+                        va_list varargs_0, va_list varargs_1);
+
+// Formats the status as a multi-line string containing all associated payloads.
+// Note that this may contain PII such as file paths and must only be used for
+// presenting errors to users and not sent to a logs aggregation service.
```

Comment:
How likely is it that `iree_status_format` will contain PII? Would a warning about including PII in messages be more appropriate than a warning about sending the strings this produces to a log sink? I think we'll find it hard to get developers to omit or filter these useful logs and we'll have more success encouraging them to think carefully about what appears in status messages (don't log "failed to open file 'my_account_information_is_foo.txt").

Reply:
Both are useful approaches, but really this depends on the application. If building a dev tool I most definitely want file paths and named identifiers (function names that threw errors, etc) in my logs. If building a release app then I'm likely stripping log messages for obfuscation/binary size/etc issues anyway and there won't be an issue.

Reply:
(in other words, it's up to the integrator to decide whether the logs are sensitive or not based on their policies)